### ISSUE TITLE
Is there a way to exclude a file from parallel execution?

### ISSUE THREAD

**Issue by trabulmonkee** [seattlerb/minitest#510]
I have a suite of many, unit style, test (Minitest 5.5.0) files that I want the full suite to run in parallel and random.  This I have working with the use of parallelize_me! at the global level.  However there are some instances where a few test files with several test cases included should not run in parallel.  How can one go about configuring the individual test file to not run in parallel, but still be included in the entire parallel execution suite?

I have tried without success, these will just run in sorted order, but will still be in parallel
* self.test_order = :alpha
* i_suck_and_my_tests_are_order_dependent!  # great method name by the way, makes me laugh. 

What I am after is a way to run the entire suite in parallel and turn off parallel execution where necessary, something similar to i_suck_and_my_tests_cannot_run_parallel!

**Comment by zenspider** [comment]
`parallelize_me!` is opt-in, not opt-out. So all subclasses of `Test` run serially by default. I'm not sure what you mean by "the use of parallelize_me! at the global level", but that sounds like the first thing you should stop. Please tell me you didn't use minitest/hell--it is named that for a reason. Then, either make a ParallelTestCase subclass that you switch most of your test classes to and call `parallelize_me!` in there, or just call it on all the classes you want to opt-in on.

Does that help? or did I miss something?

**Comment by trabulmonkee** [comment]
Nope, no minitest/hell, but why not?  My understanding was that minitest/hell does the same thing as parallelize_me!

Within the Webdriver framework I have created.  I have parallelize_me! set as opt-in as the default.  Mainly to take advantage of threading to speed up test executions on the CI server.  All test files call a helper file.  The helper sets parallelize_me! via a runtime argument to turn it off (opt-out), if need be, which helps with local debugging. The way I have it currently setup its an all or nothing option via the rake task that is executed on the CI server.

For example:
```ruby
# test_file.rb
require 'helper_file.rb'
class TS_MYTEST < Minitest::Test

  def setup
    # do setup stuff
  end

  def teardown
    # do teardown stuff
  end

  def test_something
    # webdriver test logic...
  end

  def test_something_else
     # webdriver test logic
  end
end

#helper_file.rb
require 'rubygems'
gem 'minitest'
require 'minitest/autorun'

parallel = ARGV[0]
unless parallel == 'serial'
   # this is how I believe I am achieving the parallelize_me! at the global level
   Minitest::Test.parallelize_me!
end
```
Then I can execute like this:
```ruby test_file.rb ``` to run in parallel
or
```ruby test_file.rb 'serial' ``` to run serial

99% of the time I want parallel and my foresight did not go far enough to think otherwise.  I was hoping for a solution that could opt-out at per file level, instead of refactoring, all the test files currently in place, but perhaps that is my only option based on your advice above.

Is the way I would need to go about it, as you describe, something like this?

``` ruby
class ParallelTestCase < Minitest::Test
  parallelize_me!
end

class TS_MYTEST_PARALLEL < ParallelTestCase
  def setup
    # do setup stuff
  end

  def teardown
    # do teardown stuff
  end

  def test_something
    # webdriver test logic...
  end

  def test_something_else
     # webdriver test logic
  end
end

class TS_MYTEST_SERIAL < Minitest::Test
  def setup
    # do setup stuff
  end

  def teardown
    # do teardown stuff
  end

  def test_something
    # webdriver test logic...
  end

  def test_something_else
     # webdriver test logic
  end
end
```

**Comment by zenspider** [comment]
Yup yup. That should totally work and make it obvious when you want to go parallel and when you don't. 

If you wanted, you could do what you're currently doing and then define something like:

```ruby
    def self.serialize_me!
      :random
    end
```

and then call it as opt-out.

**Comment by trabulmonkee** [comment]
I like that!   I tried it out with the current implementation, but found that even with serialize_me!, it was still getting trumped by parallelize_me!

From what I could tell once parallelize_me! is called the included files/methods of that module are in scope for the duration.

I took your advice and stopped doing it at the global level and switched to the per file implementation, via subclass.  It was a quick find/replace so not as much of headache as I had originally thought.

Thanks.
